src/dune_rules/ctypes_rules.ml

```diff
@@ -0,0 +1,624 @@
+        let path = Path.relative build_dir (String.lowercase name ^ ".ml") in
+        let impl = Module.File.make Dialect.ocaml path in
+        let source = Module.Source.make ~impl module_name in
+        Module.of_source ~visibility:Visibility.Public
+          ~kind:Module.Kind.Impl source)
+    in
+    Module.Name_map.of_list_exn modules
+  in
+  Modules.exe_unwrapped name_map
+  (* Modules.exe_wrapped ~src_dir:dir ~modules:name_map *)
+
+let write_c_types_includer_module ~sctx ~dir ~filename
+      ~type_description_functor ~c_generated_types_module =
+  let path = Path.Build.relative dir filename in
+  let contents =
+    let buf = Buffer.create 1024 in
+    let pr buf fmt = Printf.bprintf buf (fmt ^^ "\n") in
+    pr buf "include %s.Types (%s)"
+        (Module_name.to_string type_description_functor)
+        (Module_name.to_string c_generated_types_module);
+    Buffer.contents buf
+  in
+  Super_context.add_rule ~loc:Loc.none sctx ~dir
+    (Action_builder.write_file path contents)
+
+let write_entry_point_module ~ctypes ~sctx ~dir ~filename
+      ~type_description_instance ~function_description
+      ~c_types_includer_module =
+  let path = Path.Build.relative dir filename in
+  let contents =
+    let buf = Buffer.create 1024 in
+    let pr buf fmt = Printf.bprintf buf (fmt ^^ "\n") in
+    pr buf "module %s = %s"
+      (Module_name.to_string type_description_instance)
+      (Module_name.to_string c_types_includer_module);
+    List.iter function_description ~f:(fun fd ->
+      let c_generated_functions_module =
+        Stanza_util.c_generated_functions_module ctypes fd
+      in
+      pr buf "module %s = %s.Functions (%s)"
+        (fd.Ctypes.Function_description.instance |> Module_name.to_string)
+        (fd.Ctypes.Function_description.functor_ |> Module_name.to_string)
+        (Module_name.to_string c_generated_functions_module)
+    );
+    Buffer.contents buf
+  in
+  Super_context.add_rule ~loc:Loc.none sctx ~dir
+    (Action_builder.write_file path contents)
+
+let discover_gen ~external_library_name:lib ~cflags_sexp ~c_library_flags_sexp =
+  let buf = Buffer.create 1024 in
+  let pr buf fmt = Printf.bprintf buf (fmt ^^ "\n") in
+  pr buf "module C = Configurator.V1";
+  pr buf "let () =";
+  pr buf "  C.main ~name:\"%s\" (fun c ->" lib;
+  pr buf "    let default : C.Pkg_config.package_conf =";
+  pr buf "      { libs   = [\"-l%s\"];" lib;
+  pr buf "        cflags = [\"-I/usr/include\"] }";
+  pr buf "    in";
+  pr buf "    let conf =";
+  pr buf "      match C.Pkg_config.get c with";
+  pr buf "      | None -> default";
+  pr buf "      | Some pc ->";
+  pr buf "        match C.Pkg_config.query pc ~package:\"%s\" with" lib;
+  pr buf "        | None -> default";
+  pr buf "        | Some deps -> deps";
+  pr buf "    in";
+  pr buf "    C.Flags.write_sexp \"%s\" conf.cflags;" cflags_sexp;
+  pr buf "    C.Flags.write_sexp \"%s\" conf.libs;" c_library_flags_sexp;
```

Comment:
I think a better implementation would not use configurator, but rather add builtin support for pkg-config for dune itself. This way we aren't forcing to use configurator, and we don't need to generate ocaml as intermediate step to run a binary.